### PR TITLE
Import a single Trello card (and its comments)

### PR DESCRIPTION
**Pull request by xavierpriour**
This is step 1 of #113 
We import a single card, with its title, description, comments, and labels.
Limitations (to be fixed in future PRs):

* card members are removed
* Trello history is lost
* attachments are not transferred

### CODE REVIEW

**Comment by xavierpriour** [comment]
Oops forgot to eslint - will fix!

**Comment by xavierpriour** [comment]
merged with core wekan/devel + fixed an eslint issue in one of the new files - now passing :)

File: client/components/lists/listHeader.jade [hunk 1]
```diff
@@ -25,11 +25,20 @@ template(name="listActionPopup")
...
     li: a.js-close-list {{_ 'archive-list'}}
 
 template(name="listMoveCardsPopup")
   +boardLists
 
+template(name="listImportCardPopup")
+  form
+    label
+      | {{_ 'card-json'}}
+      //+editor(class="js-card-json" autofocus=true)
```
Comment (mquandalle) [inline client/components/lists/listHeader.jade]: to remove?

File: client/components/lists/listHeader.js [hunk 2]
```diff
@@ -47,6 +49,70 @@ Template.listActionPopup.events({
   },
 });
 
+Template.listImportCardPopup.events({
+  submit(evt) {
+    // 1. get the json data out of the form and parse it
+    evt.preventDefault();
+    const jsonData = $(evt.currentTarget).find('textarea').val();
+    const data = JSON.parse(jsonData);
```
Comment (mquandalle) [inline client/components/lists/listHeader.js]: I think it’s worth implementing error handling here before merging the feature. Maybe we could use something like [this](https://github.com/wekan/wekan/blob/758be3356cc3fb05de0a29787059b6b34aac08a2/client/components/users/userAvatar.js#L100-L106)?

File: client/components/lists/listHeader.js [hunk 3]
```diff
@@ -47,6 +49,70 @@ Template.listActionPopup.events({
...
+    const cardToCreate = {
+      title: data.name,
+      description: data.desc,
+      listId: this._id,
+      boardId: this.boardId,
+      userId: Meteor.userId(),
+      sort: sortIndex,
+    };
+    // 3. map labels
+    data.labels.forEach((current) => {
```
Comment (mquandalle) [inline client/components/lists/listHeader.js]: `currentLabel`?

File: client/lib/accessibility.js [hunk 4]
```diff
@@ -7,7 +7,7 @@
 // Without an href, links are non-keyboard-focusable and are not presented on
 // blind screen readers. We default to the empty anchor `#` href.
 function enforceHref(attributes) {
-  if (! _.has(attributes, 'href')) {
+  if (!_.has(attributes, 'href')) {
```
Comment (mquandalle) [inline client/lib/accessibility.js]: Sorry for that, I had disabled ESLint on my side :-(

**Comment by mquandalle** [comment]
Hi Xavier,

Thank you for the PR! The code looks good to me. Apart from my tiny inline comments there is one bigger change I would like to consider with you before merging this PR. I don’t know how much you are familiar with Meteor, but basically there are two paths to mutate the database, one is using `Meteor.methods` that are basically RPC (but with potential optimistic update on the client) and the other is using minimongo operations (like we everywhere in Wekan). The thing is that with minimongo `.insert` we enforce some constants, for instance we ensure that the `createdAt` date is correct ([here](https://github.com/wekan/wekan/blob/758be3356cc3fb05de0a29787059b6b34aac08a2/models/cards.js#L195)) but in the case of the card import it is probably better to trust the data we are importing (or at least ensure a set of different rules that we would normally do). So I propose to do the import in a Meteor Method, because it makes it easy to chose which validation rule we want to enforce.

Does that makes sense?

File: client/components/lists/listHeader.jade (hunk 1 above)
Comment (xavierpriour) [inline client/components/lists/listHeader.jade]: sorry!

File: i18n/en.i18n.json [hunk 5]
```diff
@@ -65,6 +65,8 @@
     "card-edit-attachments": "Edit attachments",
     "card-edit-labels": "Edit labels",
     "card-edit-members": "Edit members",
+    "card-json": "Go to a Trello card, select 'Share and more...' then 'Export JSON' and copy the resulting text",
```
Comment (mquandalle) [inline i18n/en.i18n.json]: [..] copy **and paste** the [..]?

File: client/components/lists/listHeader.jade (hunk 1 above)
Comment (mquandalle) [inline client/components/lists/listHeader.jade]: Don’t be sorry :-)

**Comment by xavierpriour** [comment]
Perfect sense, and addresses some of my concerns mentioned in #113 
What file should I put my method definition in, though?

**Comment by mquandalle** [comment]
I guess `models/cards.js` — but we’ll probably want to split this file later on.

**Comment by xavierpriour** [comment]
Well, I'm going to start work on importing boards next - so mind if I create a models/trello.js right now and put the method in it?

**Comment by mquandalle** [comment]
Ok, but name if `import.js` because at some point we’ll be able to import data exported from Wekan as well :-)

**Comment by xavierpriour** [comment]
Sure!

**Comment by mquandalle** [comment]
(Or even #152 :-))

File: client/components/lists/listHeader.js (hunk 2 above)
Comment (xavierpriour) [inline client/components/lists/listHeader.js]: I'm having trouble accessing the `error` field (this.error is undefined in my method) - where is it defined?

Comment (mquandalle) [inline client/components/lists/listHeader.js]: [here](https://github.com/wekan/wekan/blob/758be3356cc3fb05de0a29787059b6b34aac08a2/client/components/users/userAvatar.js#L56-L58)

**Comment by xavierpriour** [comment]
When generating activities for the card import itself, should it generate an 'importCard' only, or an 'importCard' + a 'createCard' (with same date)?

**Comment by mquandalle** [comment]
I think only “importCard” will be less confusing.

File: models/import.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
+Meteor.methods({
+  /**
+   *
+   */
+  importTrelloCard(trelloCard, listId, sortIndex) {
+    // 1. check parameters are ok from a syntax point of view
```
Comment (mquandalle) [inline models/import.js]: :+1:

File: models/import.js [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+      })],
+      members: [Object],
+    }));
+    check(listId, String);
+    check(sortIndex, Number);
+
+    // 2. check parameters are ok from a business point of view (exist & authorized)
+    const list = Lists.findOne(listId);
+    if(!list) {
+      throw 'exception-list-doesNotExist';
```
Comment (mquandalle) [inline models/import.js]: I think the convention with Meteor is to throw a `Meteor.Error`

Comment (xavierpriour) [inline models/import.js]: Yes, I'm still on it - I pushed a little too early :)

Comment (mquandalle) [inline models/import.js]: That’s fine, I close my eyes :-)

**Comment by xavierpriour** [comment]
@mquandalle it should be good now

File: client/components/lists/listHeader.js [hunk 8]
```diff
@@ -47,6 +49,45 @@ Template.listActionPopup.events({
   },
 });
 
+
+BlazeComponent.extendComponent({
+  events() {
+    return [{
+      'submit': (evt) => {
```
Comment (mquandalle) [inline client/components/lists/listHeader.js]: Not important, but our eslint convention is `submit(evt) {`

Comment (xavierpriour) [inline client/components/lists/listHeader.js]: Done!

**Comment by mquandalle** [comment]
Tested and it works well! Thank you Xavier, I’m ready to merge. Maybe could you just `git rebase -i  21dcd16` and squashes some of the commits (for instance the “fix eslint style”) so we get a cleaner git history?

**Comment by xavierpriour** [comment]
ok, did some squashing.

**Comment by xavierpriour** [comment]
ok, something seems wrong, rebase brought your commits in my PR... how do I get them out?

**Comment by mquandalle** [comment]
Yep something went wrong, I'll fix and merge.

**Comment by mquandalle** [comment]
Merged in b670a1ab36fcf514751e4b356487a96088dbbd24, thanks!

**Pull request closed without merging**